---
title: "Class 16: Pathway Analysis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Live code: 
See the live code version here 

https://tinyurl.com/class16-bggn213 

## Download and read input data 
The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. 



```{r}
# Import metadata and take a peak
colData = read.csv("GSE37704_metadata.csv", row.names=1)
head(colData)
```

```{r}
# Import countdata
rawcountData = read.csv("GSE37704_featurecounts.csv", row.names=1)
head(countData)
```

```{r}
countData <- rawcountData[,-1]
head(countData)
```

```{r}
dim(countData)
```


There are genes that have zero counts in all samples (i.e. columns). We should remove these. 

First identify the zero sum rows and then Exclude these zero sum rows 
```{r}
countData <- countData [rowSums(countData) != 0,]


```

```{r}
dim(countData)
```

## Running DESeq2
```{r}
library(DESeq2)
```


```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
res <- results(dds)
head(res)
```

```{r}
plot(res$log2FoldChange, -log(res$padj))
```

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "blue"


plot( res$log2FoldChange, -log(res$padj), col=mycols)
```

To add annotation to my gene list I will install some bioconductor packages
However, i do nott do this in my document I install them in the CONSOLE!!

```{r}
#BiocManager::install("AnnotationDbi")
#BiocManager::install("org.Hs.eg.db")
```


```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

columns(org.Hs.eg.db)
```

Let's map our ENSEMBLE gene ids to the more conventional gene SYMBOL

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys = row.names(res), 
                     keytype="ENSEMBL",
                     column="SYMBOL",
                     multiVals="first")

res$entrez <- mapIds(org.Hs.eg.db,
                     keys = row.names(res), 
                     keytype="ENSEMBL",
                     column="ENTREZID",
                     multiVals="first")
  
res$name <- mapIds(org.Hs.eg.db,
                     keys = row.names(res), 
                     keytype="ENSEMBL",
                     column="GENENAME",
                     multiVals="first")
  
head(res)
```

Save our annotated results

```{r}
write.csv(res, file="deseq_myresults.csv")
```



# Pathway analysis

Here we are going to use the **gage** package for pathway analysis. Once we have a list of enriched pathways, we're going to use the **pathview** package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
# Run in your R console (i.e. not your Rmarkdown doc!)
#BiocManager::install( c("pathview", "gage", "gageData") )
```

Load these packages

```{r}
library(gage)
library(gageData)
library(pathview)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Our input will be a vector of fold change values with ENTREZ gene ids as names
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id = "hsa04110")
```

```{r}
knitr::include_graphics("hsa04110.pathview.png")
```


Try for high resolution figure
```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

To use reactome online we need a files of gene IDs in a particular format 
```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```




